File: clients/fsharp/Model/Unit.fs
#nowarn "0058"

namespace AiCup22
namespace AiCup22.Model

open AiCup22

/// A unit
type Unit = {
    /// Unique id
    Id: int;
    /// Id of the player (team) controlling the unit
    PlayerId: int;
    /// Current health
    Health: double;
    /// Current shield value
    Shield: double;
    /// Left extra lives of this unit
    ExtraLives: int;
    /// Current position of unit's center
    Position: Model.Vec2;
    /// Remaining time until unit will be spawned, or None
    RemainingSpawnTime: option<double>;
    /// Current velocity
    Velocity: Model.Vec2;
    /// Current view direction (vector of length 1)
    Direction: Model.Vec2;
    /// Value describing process of aiming (0 - not aiming, 1 - ready to shoot)
    Aim: double;
    /// Current action unit is performing, or None
    Action: option<Model.Action>;
    /// Tick when health regeneration will start (can be less than current game tick)
    HealthRegenerationStartTick: int;
    /// Index of the weapon this unit is holding (starting with 0), or None
    Weapon: option<int>;
    /// Next tick when unit can shoot again (can be less than current game tick)
    NextShotTick: int;
    /// List of ammo in unit's inventory for every weapon type
    Ammo: int[];
    /// Number of shield potions in inventory
    ShieldPotions: int;
} with

    /// Write Unit to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.Id
        writer.Write this.PlayerId
        writer.Write this.Health
        writer.Write this.Shield
        writer.Write this.ExtraLives
        this.Position.writeTo writer
        match this.RemainingSpawnTime with
            | Some value ->
                writer.Write true
                writer.Write value
            | None -> writer.Write false
        this.Velocity.writeTo writer
        this.Direction.writeTo writer
        writer.Write this.Aim
        match this.Action with
            | Some value ->
                writer.Write true
                value.writeTo writer
            | None -> writer.Write false
        writer.Write this.HealthRegenerationStartTick
        match this.Weapon with
            | Some value ->
                writer.Write true
                writer.Write value
            | None -> writer.Write false
        writer.Write this.NextShotTick
        writer.Write this.Ammo.Length
        this.Ammo |> Array.iter (fun value ->
            writer.Write value )
        writer.Write this.ShieldPotions
        ()

    /// Read Unit from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Id = reader.ReadInt32()
        PlayerId = reader.ReadInt32()
        Health = reader.ReadDouble()
        Shield = reader.ReadDouble()
        ExtraLives = reader.ReadInt32()
        Position = Model.Vec2.readFrom reader;
        RemainingSpawnTime = match reader.ReadBoolean() with
                                 | true -> Some(reader.ReadDouble())
                                 | false -> None
        Velocity = Model.Vec2.readFrom reader;
        Direction = Model.Vec2.readFrom reader;
        Aim = reader.ReadDouble()
        Action = match reader.ReadBoolean() with
                     | true -> Some(Model.Action.readFrom reader;)
                     | false -> None
        HealthRegenerationStartTick = reader.ReadInt32()
        Weapon = match reader.ReadBoolean() with
                     | true -> Some(reader.ReadInt32())
                     | false -> None
        NextShotTick = reader.ReadInt32()
        Ammo = [|for _ in 1 .. reader.ReadInt32() do
                   yield reader.ReadInt32() |]
        ShieldPotions = reader.ReadInt32()
    }

File: clients/csharp/Model/Unit.cs
namespace AiCup22.Model
{
    /// <summary>
    /// A unit
    /// </summary>
    public struct Unit
    {
        /// <summary>
        /// Unique id
        /// </summary>
        public int Id { get; set; }
        /// <summary>
        /// Id of the player (team) controlling the unit
        /// </summary>
        public int PlayerId { get; set; }
        /// <summary>
        /// Current health
        /// </summary>
        public double Health { get; set; }
        /// <summary>
        /// Current shield value
        /// </summary>
        public double Shield { get; set; }
        /// <summary>
        /// Left extra lives of this unit
        /// </summary>
        public int ExtraLives { get; set; }
        /// <summary>
        /// Current position of unit's center
        /// </summary>
        public AiCup22.Model.Vec2 Position { get; set; }
        /// <summary>
        /// Remaining time until unit will be spawned, or None
        /// </summary>
        public double? RemainingSpawnTime { get; set; }
        /// <summary>
        /// Current velocity
        /// </summary>
        public AiCup22.Model.Vec2 Velocity { get; set; }
        /// <summary>
        /// Current view direction (vector of length 1)
        /// </summary>
        public AiCup22.Model.Vec2 Direction { get; set; }
        /// <summary>
        /// Value describing process of aiming (0 - not aiming, 1 - ready to shoot)
        /// </summary>
        public double Aim { get; set; }
        /// <summary>
        /// Current action unit is performing, or None
        /// </summary>
        public AiCup22.Model.Action? Action { get; set; }
        /// <summary>
        /// Tick when health regeneration will start (can be less than current game tick)
        /// </summary>
        public int HealthRegenerationStartTick { get; set; }
        /// <summary>
        /// Index of the weapon this unit is holding (starting with 0), or None
        /// </summary>
        public int? Weapon { get; set; }
        /// <summary>
        /// Next tick when unit can shoot again (can be less than current game tick)
        /// </summary>
        public int NextShotTick { get; set; }
        /// <summary>
        /// List of ammo in unit's inventory for every weapon type
        /// </summary>
        public int[] Ammo { get; set; }
        /// <summary>
        /// Number of shield potions in inventory
        /// </summary>
        public int ShieldPotions { get; set; }
    
        public Unit(int id, int playerId, double health, double shield, int extraLives, AiCup22.Model.Vec2 position, double? remainingSpawnTime, AiCup22.Model.Vec2 velocity, AiCup22.Model.Vec2 direction, double aim, AiCup22.Model.Action? action, int healthRegenerationStartTick, int? weapon, int nextShotTick, int[] ammo, int shieldPotions)
        {
            this.Id = id;
            this.PlayerId = playerId;
            this.Health = health;
            this.Shield = shield;
            this.ExtraLives = extraLives;
            this.Position = position;
            this.RemainingSpawnTime = remainingSpawnTime;
            this.Velocity = velocity;
            this.Direction = direction;
            this.Aim = aim;
            this.Action = action;
            this.HealthRegenerationStartTick = healthRegenerationStartTick;
            this.Weapon = weapon;
            this.NextShotTick = nextShotTick;
            this.Ammo = ammo;
            this.ShieldPotions = shieldPotions;
        }
    
        /// <summary> Read Unit from reader </summary>
        public static Unit ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new Unit();
            result.Id = reader.ReadInt32();
            result.PlayerId = reader.ReadInt32();
            result.Health = reader.ReadDouble();
            result.Shield = reader.ReadDouble();
            result.ExtraLives = reader.ReadInt32();
            result.Position = AiCup22.Model.Vec2.ReadFrom(reader);
            if (reader.ReadBoolean())
            {
                result.RemainingSpawnTime = reader.ReadDouble();
            } else
            {
                result.RemainingSpawnTime = null;
            }
            result.Velocity = AiCup22.Model.Vec2.ReadFrom(reader);
            result.Direction = AiCup22.Model.Vec2.ReadFrom(reader);
            result.Aim = reader.ReadDouble();
            if (reader.ReadBoolean())
            {
                result.Action = AiCup22.Model.Action.ReadFrom(reader);
            } else
            {
                result.Action = null;
            }
            result.HealthRegenerationStartTick = reader.ReadInt32();
            if (reader.ReadBoolean())
            {
                result.Weapon = reader.ReadInt32();
            } else
            {
                result.Weapon = null;
            }
            result.NextShotTick = reader.ReadInt32();
            result.Ammo = new int[reader.ReadInt32()];
            for (int ammoIndex = 0; ammoIndex < result.Ammo.Length; ammoIndex++)
            {
                result.Ammo[ammoIndex] = reader.ReadInt32();
            }
            result.ShieldPotions = reader.ReadInt32();
            return result;
        }
    
        /// <summary> Write Unit to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(Id);
            writer.Write(PlayerId);
            writer.Write(Health);
            writer.Write(Shield);
            writer.Write(ExtraLives);
            Position.WriteTo(writer);
            if (!RemainingSpawnTime.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                writer.Write(RemainingSpawnTime.Value);
            }
            Velocity.WriteTo(writer);
            Direction.WriteTo(writer);
            writer.Write(Aim);
            if (!Action.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                Action.Value.WriteTo(writer);
            }
            writer.Write(HealthRegenerationStartTick);
            if (!Weapon.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                writer.Write(Weapon.Value);
            }
            writer.Write(NextShotTick);
            writer.Write(Ammo.Length);
            foreach (var ammoElement in Ammo)
            {
                writer.Write(ammoElement);
            }
            writer.Write(ShieldPotions);
        }
    
        /// <summary> Get string representation of Unit </summary>
        public override string ToString() {
            string stringResult = "Unit { ";
            stringResult += "Id: ";
            stringResult += Id.ToString();
            stringResult += ", ";
            stringResult += "PlayerId: ";
            stringResult += PlayerId.ToString();
            stringResult += ", ";
            stringResult += "Health: ";
            stringResult += Health.ToString();
            stringResult += ", ";
            stringResult += "Shield: ";
            stringResult += Shield.ToString();
            stringResult += ", ";
            stringResult += "ExtraLives: ";
            stringResult += ExtraLives.ToString();
            stringResult += ", ";
            stringResult += "Position: ";
            stringResult += Position.ToString();
            stringResult += ", ";
            stringResult += "RemainingSpawnTime: ";
            if (!RemainingSpawnTime.HasValue)
            {
                stringResult += "null";
            } else
            {
                stringResult += RemainingSpawnTime.Value.ToString();
            }
            stringResult += ", ";
            stringResult += "Velocity: ";
            stringResult += Velocity.ToString();
            stringResult += ", ";
            stringResult += "Direction: ";
            stringResult += Direction.ToString();
            stringResult += ", ";
            stringResult += "Aim: ";
            stringResult += Aim.ToString();
            stringResult += ", ";
            stringResult += "Action: ";
            if (!Action.HasValue)
            {
                stringResult += "null";
            } else
            {
                stringResult += Action.Value.ToString();
            }
            stringResult += ", ";
            stringResult += "HealthRegenerationStartTick: ";
            stringResult += HealthRegenerationStartTick.ToString();
            stringResult += ", ";
            stringResult += "Weapon: ";
            if (!Weapon.HasValue)
            {
                stringResult += "null";
            } else
            {
                stringResult += Weapon.Value.ToString();
            }
            stringResult += ", ";
            stringResult += "NextShotTick: ";
            stringResult += NextShotTick.ToString();
            stringResult += ", ";
            stringResult += "Ammo: ";
            stringResult += "[ ";
            int ammoIndex = 0;
            foreach (var ammoElement in Ammo)
            {
                if (ammoIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += ammoElement.ToString();
                ammoIndex++;
            }
            stringResult += " ]";
            stringResult += ", ";
            stringResult += "ShieldPotions: ";
            stringResult += ShieldPotions.ToString();
            stringResult += " }";
            return stringResult;
        }
    }
}